src/mpi/coll/src/csel.c

```diff
@@ -0,0 +1,1253 @@
+        if (node->failure) {
+            fprintf(stderr, "unexpected non-NULL failure path for coll %d\n", coll);
+            MPIR_Assert(0);
+        }
+    } else if (node->type != CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTRA &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTER &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_PARENT_COMM &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_NODE_CONSECUTIVE &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_SBUF_INPLACE &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_BLOCK_REGULAR &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_COMMUTATIVE &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__IS_OP_BUILT_IN &&
+               node->type != CSEL_NODE_TYPE__OPERATOR__COLLECTIVE) {
+        /* for boolean types, the failure path might or might not be
+         * NULL, but for everything else, the failure path must not be
+         * NULL */
+        if (node->failure == NULL) {
+            fprintf(stderr, "unexpected NULL failure path for coll %d\n", coll);
+            MPIR_Assert(0);
+        }
+    }
+
+    if (node->success)
+        validate_tree(node->success);
+    if (node->failure)
+        validate_tree(node->failure);
+}
+
+static csel_node_s *parse_json_tree(struct json_object *obj,
+                                    void *(*create_container) (struct json_object *))
+{
+    enum json_type type;
+    csel_node_s *prevnode, *tmp, *node = NULL;
+
+    json_object_object_foreach(obj, key, val) {
+        type = json_object_get_type(val);
+        MPIR_Assert(type == json_type_object);
+
+        char *ckey = MPL_strdup_no_spaces(key);
+
+        tmp = MPL_malloc(sizeof(csel_node_s), MPL_MEM_COLL);
+
+        if (!strncmp(ckey, "composition=", strlen("composition=")) ||
+            !strncmp(ckey, "algorithm=", strlen("algorithm="))) {
+            tmp->type = CSEL_NODE_TYPE__CONTAINER;
+            tmp->u.cnt.container = create_container(obj);
+            MPL_free(ckey);
+            return tmp;
+        }
+
+        /* this node must be an operator type */
+        tmp->success = parse_json_tree(json_object_object_get(obj, key), create_container);
+        tmp->failure = NULL;
+
+        if (node == NULL)
+            node = tmp;
+        else if (prevnode != NULL)
+            prevnode->failure = tmp;
+        prevnode = tmp;
+
+        if (!strcmp(ckey, "comm_type=intra")) {
+            tmp->type = CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTRA;
+        } else if (!strcmp(ckey, "comm_type=inter")) {
+            tmp->type = CSEL_NODE_TYPE__OPERATOR__COMM_TYPE_INTER;
+        } else if (!strncmp(ckey, "collective=", strlen("collective="))) {
+            tmp->type = CSEL_NODE_TYPE__OPERATOR__COLLECTIVE;
+
+            char *str = ckey + strlen("collective=");
+            if (!strcmp(str, "allgather"))
+                tmp->u.collective.coll_type = MPIR_CSEL_COLL_TYPE__ALLGATHER;
```

Comment:
Suggest refactor with a routine, e.g. `coll_type_str_to_id`

Reply:
Unless there's an automatic way to do this (and I don't mean with a script), it's the same amount of code to be maintained.